### PR TITLE
Validate monitoring header overrides at parse time

### PR DESCRIPTION
**Pull request by danhermann**
Provides parse-time validation for `HEADERS_SETTING` as described in #47711.

~~Also changes the check for blacklisted header overrides to be [case-insensitive](https://tools.ietf.org/html/rfc7230#section-3.2).~~

### CODE REVIEW

**Review by jakelandis — APPROVED** [review]
LGTM assuming there are pre-existing tests that cover this.

**Comment by danhermann** [comment]
Thanks, @jakelandis, there are existing tests for this setting.

**Comment by danhermann** [comment]
@elastic/es-core-infra, this work covers both monitoring and settings if you want to review it.

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java [hunk 1]
```diff
@@ -209,7 +209,23 @@ public void validate(final List<String> hosts, final Map<Setting<?>, Object> set
...
     public static final Setting.AffixSetting<Settings> HEADERS_SETTING =
             Setting.affixKeySetting("xpack.monitoring.exporters.","headers",
-                    (key) -> Setting.groupSetting(key + ".", Property.Dynamic, Property.NodeScope));
+                    (key) -> Setting.groupSetting(
+                        key + ".",
+                        settings -> {
+                            final Set<String> names = settings.names();
+                            for (String name : names) {
+                                final String fullSetting = key + "." + name;
+                                if (HttpExporter.BLACKLISTED_HEADERS.stream().anyMatch(name::equalsIgnoreCase)) {
```
Comment (rjernst) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java]: Why use equalsIgnoreCase? I don't see where we've been case insensitive in setting names before, and I'm not sure we should change behavior here while trying to add validation.

**Review by rjernst — CHANGES_REQUESTED** [review]
One comment

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java (hunk 1 above)
Comment (danhermann) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java]: HTTP headers are supposed to be case insensitive, so without the case insensitive check, a user could supply an override for "content-length" even though "Content-Length" is on the header override blacklist. If that's not a concern here or this is the wrong place to check that, I can make it a case sensitive check.

Comment (danhermann) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java]: @rjernst, would you prefer the HTTP header name check here remain case-sensitive?

Comment (rjernst) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java]: I'm open to it being case-insensitive, but changing that behavior should be separate PR. This PR is just about adding validation to the parsing as it currently exists.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine update branch

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java (hunk 1 above)
Comment (danhermann) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExporter.java]: @rjernst, I've restored the original case-sensitive check here.

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**